packages/editor/src/systems/RenderMonitorSystem.tsx

```diff
@@ -0,0 +1,130 @@
+    SceneComplexityWeights.drawCallsWeight * params.drawCalls +
+    SceneComplexityWeights.shaderComplexityWeight * params.shaderComplexity
+  return complexity
+}
+
+// count number of lines in vertex and fragment shaders
+function getShaderComplexity(resources: Record<string, any>): number {
+  const countLines = (code?: string): number => (code || '').split('\n').length
+  const totalComplexity = Object.values(resources)
+    .filter((resource) => resource.type === ResourceType.Material)
+    .flatMap((resource) => (Array.isArray(resource.asset) ? resource.asset : [resource.asset]))
+    .filter((material) => material.isShaderMaterial || material.isMeshStandardMaterial)
+    .reduce(
+      (total, material) => {
+        total.vertexInstructions += countLines(material.vertexShader)
+        total.fragmentInstructions += countLines(material.fragmentShader)
+        return total
+      },
+      { vertexInstructions: 0, fragmentInstructions: 0 }
+    )
+
+  return totalComplexity.vertexInstructions + totalComplexity.fragmentInstructions
+}
+
+export const RenderMonitorSystem = defineSystem({
+  uuid: 'ee.editor.RenderMonitorSystem',
+  insert: { after: PresentationSystemGroup },
+  reactor: () => {
+    const { t } = useTranslation()
+
+    const sceneComplexityScore = useHookstate(0)
+
+    const renderInfoState = useHookstate(getMutableState(RenderInfoState))
+    const resourceState = useHookstate(getMutableState(ResourceState))
+    const lightQuery = useQuery([LightTagComponent, VisibleComponent, SourceComponent])
+
+    useEffect(() => {
+      const params = {
+        vertices: resourceState.totalVertexCount.value,
+        triangles: renderInfoState.info.triangles.value,
+        texturesMB: resourceState.totalBufferCount.value / (1024 * 1024),
+        drawCalls: renderInfoState.info.calls.value,
+        shaderComplexity: getShaderComplexity(resourceState.resources.value),
+        lights: lightQuery.length
+      }
+
+      sceneComplexityScore.set(calculateSceneComplexity(params))
+    }, [
+      resourceState.totalVertexCount,
+      resourceState.totalBufferCount,
+      renderInfoState.info.triangles.value,
+      renderInfoState.info.calls.value,
+      renderInfoState.info.programs.value,
+      lightQuery
+    ])
+
+    useEffect(() => {
+      // these thresholds are to be adjusted  based on experimentation
+      let warning = t('editor:warnings.sceneComplexity', { sceneComplexity: SceneComplexity.VeryHeavy.label })
+      if (sceneComplexityScore.value < SceneComplexity.VeryLight.value) return
+      if (sceneComplexityScore.value < SceneComplexity.Light.value) return
+      if (sceneComplexityScore.value < SceneComplexity.Medium.value) return
+      if (sceneComplexityScore.value < SceneComplexity.Heavy.value)
+        warning = t('editor:warnings.sceneComplexity', { sceneComplexity: SceneComplexity.Heavy.label })
+      getMutableState(EditorWarningState).warning.set(warning)
```

Comment:
the state we set here is a simple string. we can instead use `NotificationService.dispatchNotify` here directly. This will avoid us having `useEffect`s

Reply:
i agree, i decided to build out the warning system to allow us to dispatch warnings from anywhere in the code base more easily